esteid/idcard/README.md

```diff
@@ -30,21 +32,62 @@
+#### Compatibility
 
 Since the previous service-based action implementation provided for a very long inheritance tree
 of action classes which all override the `do_action` method in some way but still make a call to a
 non-existent `service`, it's advised that all those "actions" be scrapped in favor of a simpler API. 
 
-## The suggested API
+## Authentication with ID Card
+
+### Process Outline
+
+The request to the client to enter PIN is initialized by a specifically configured web server,
+with an option similar to `ssl_verify_client` on nginx.
+
+On nginx, it is only possible to protect an entire `server` section with this option, which means that
+in order to keep the site generally accessible to unauthenticated clients we need to set up a separate domain
+that would handle the ID card authentication process, and pass the client authentication data to the primary site
+via session or other means.
+
+There is a caveat: browsers keep a certificate in cache for an unspecified period of time,
+and a new request for PIN may not be triggered even once the authenticated ID card is removed or replaced.
+This is not possible to control by the web server. 
+
+Let's describe the key points of the ID card authentication process, based on the usual routine common for
+SmartID and MobileID. 
+
+The process thus should consist of the following steps:
+* While visiting the site (assuming `example.com` for this matter), the user clicks on a link to start the authentication process;
+* Instead of issuing a _start_ request to the backend (as in the usual routine), 
+  or a `hwcrypto` library call (as in the process of signing with ID card), the user is taken to a specifically configured
+  domain (e.g. `auth.example.com`) where they are asked for the ID Card's PIN code and 
+  proceed to allow the site to access the certificate;
+* the certificate is validated via OCSP;
+* the certificate or data obtained from it is saved to the session;
+* user is redirected to the protected page on the `example.com` site.
+
+The process above is by and large impractical. One improvement that could be made to it is to use an iframe instead of
+a redirect; this provides for a smooth procedure, but additional measures must be taken, such as adding appropriate
+`Content-Security-Policy` headers.
+
+### Notes
+
+* With an AJAX / `fetch` request, the browser (most likely the `chrome-token-signing` plugin) is not able to process
+  the response from the `auth.DOMAIN` server and shows an alert "Technical error".
+
+* There is apparently no way to access content inside the iframe from the parent window, 
+  even with CORS headers set. The only feasible way is to send `postMessage` to the parent window.
 
-TBD
+* The ID card authentication certificate is cached by the browser, 
+  and there is apparently no way to manage or work around this cache from the application.
+  The multi-domain approach (i.e. each new authentication happens on a different domain) didn't help.
```

Comment:
> The multi-domain approach (i.e. each new authentication happens on a different domain) didn't help.

Just curious, has something changed or why did it not help? Previously when the subdomain was truly just a random hash then browsers just asked for the pin again (when the card in the reader changed).

Thinking that maybe the behavior changed with the fix for the recent idcard security issue.

Reply:
https://www.id.ee/artikkel/ria-koos-partneritega-parandas-id-kaardi-brauserilaienduses-kriitilise-vea/

Reply:
Can't say if the news has anything to do with that (I rather felt it was about a possibility to pass on the info received from the ID card by e.g. nginx, to use it to access a 3rd service e.g. eesti.ee) but I started with trying the subdomain model (I remember we discussed it in slack) and whenever I got authenticated on one subdomain, I was also authenticated on others. I was using firefox because in Chrome this plugin just doesn't see the ID card. Since it's some opaque mechanism I couldn't debug, I can't really suggest any solution other than just get along with it for now. Maybe it could be found out during "production testing" ...
I would say this whole thing is a bit unreliable, as sometimes it asks me to select certificate, sometimes it doesn't, and even if I take the ID card out of the reader it kind of doesn't notice it and I keep being authenticated.

Reply:
Actually you're right, I gave it another round of tests and it seems to be working with different subdomains now. Might have been some other problem such as misconfiguration... Sorry for that.

I'm happy to make the necessary changes, but _for local testing_ there is a conundrum - if every auth frame is opened with a new domain, then before it is usable, one needs to add a security exception for that new domain (due to self-signed cert), or else the frame is blocked. So had to add a bit of guidance through this, and - another fun thing - this doesn't work in Firefox because it asks the ID card PIN *before* asking for a security exception, and this means that the frame itself will not ask the ID card credentials (at least for the period of caching). Luckily, for some reason the plugin started working in Chrome (maybe a version update helped) so it feels this concept is usable.

Again, my apologies for not being thorough from the start.